### PR TITLE
[dune] [ci] Fix and test release profile + use 1.1 dune-workspace

### PR DESCRIPTION
**Pull request by ejgallego**
Dune 1.1 allows us to define the `env` flags in the workspace file,
which is a better place than the current situation.

Along the way, We fix the build flags for release mode [missing
`-rectypes` and add a `release` build profile CI job.

### CODE REVIEW

File: .gitlab-ci.yml
Comment (Zimmi48) [inline .gitlab-ci.yml]: The number of different build jobs is becoming a bit too much :(

File: dune-workspace
Comment (Zimmi48) [inline dune-workspace]: I don't get it. We are silencing some warnings in dev mode but not in release mode? Why would we want users to see warnings that we don't see?

**Comment by Zimmi48** [comment]
If #8411 is merged first, then this PR will break it. If you want to proceed and merge this, #8411 would have to wait that Dune is updated to 1.1 in nixpkgs.

**Comment by gares** [comment]
It is not fully in topic but not off topic either: in another PR it was pointed out that dune natively supports a "profile" that uses opaque in order to speed up re-compilation. Could we have that too? Is it already there?

**Review by gares — APPROVED** [review]
for what I understand, this it ok

File: .gitlab-ci.yml
Comment (ejgallego) [reply .gitlab-ci.yml]: This would be the seed for the OPAM packaging job, it could be moved along the `pkg:nix` job tho.

File: dune-workspace
Comment (ejgallego) [reply dune-workspace]: The `env` profile uses a much stricter set of warnings than the `release` one.

**Comment by ejgallego** [comment]
> It is not fully in topic but not off topic either: in another PR it was pointed out that dune natively supports a "profile" that uses opaque in order to speed up re-compilation. Could we have that too? Is it already there?

This the built-in `dev` profile which is the default if you don't use `-p` .

File: .gitlab-ci.yml
Comment (ejgallego) [reply .gitlab-ci.yml]: In fact the system is ready to generate the OPAM packages now, but I was planning to wait to:
- add documentation build target
- add test-suite build target
- fix the couple of bugs around [such as coqide path detection]
- wait for the better version number system to land in Dune.

Comment (Zimmi48) [reply .gitlab-ci.yml]: Fine by me to do this in several steps. But indeed, do move this next to the pkg:nix job. If fact, once the documentation / test-suite targets are ready, we could even use this to build the Nix package.

Comment (ejgallego) [reply .gitlab-ci.yml]: Ok, moving to `pkg:dune-release`

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml]: When putting stuff in the test stage we should use `dependencies: []` to avoid downloading the various built coqs.

Comment (SkySkimmer) [reply .gitlab-ci.yml]: In fact let's put it in the dune-template as that seems simpler.

Comment (ejgallego) [reply .gitlab-ci.yml]: Good point, thanks.

File: Makefile.dune
Comment (SkySkimmer) [inline Makefile.dune]: Do we really want to align stuff here? It ends up making updates annoying.

Comment (ejgallego) [reply Makefile.dune]: Given that this is a document I consider formatting important. The alternative would be to use a table generator.

File: dune-workspace
Comment (SkySkimmer) [reply dune-workspace]: I don't want to look up the dune defaults to figure out which warnings are on, can we do like in configure and use `+a...`?

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml]: What's this `pkg` prefix?

File: dune-workspace
Comment (ejgallego) [reply dune-workspace]: We could do our own profile indeed, but here I think I do prefer relying on Dune's. It is one thing less to worry about, and profiles can be always be printed easily.

File: .gitlab-ci.yml
Comment (ejgallego) [reply .gitlab-ci.yml]: We took that prefix for package manager dependent builds. We have NiX, then this one in "release" will build an opam package (and maybe upload it).

An important consideration of the release mode tho is that is uses stronger optimizations, so maybe the flambda version should be built like that. I dunno.l, lots of room here.

**Review by SkySkimmer — CHANGES_REQUESTED** [review]
Request fix for the `dependencies` issue.

@Zimmi48 @ejgallego after that do I merge or do I wait for nix stuff to happen?

File: .gitlab-ci.yml
Comment (ejgallego) [reply .gitlab-ci.yml]: > In fact let's put it in the dune-template as that seems simpler.

I've leave it like this, but note that Dune has no problem in reusing previous builds in an incremental way.

**Comment by Zimmi48** [comment]
@SkySkimmer You can proceed with merging.

File: .gitlab-ci.yml
Comment (SkySkimmer) [reply .gitlab-ci.yml]: >I've left it like this, but note that Dune has no problem in reusing previous builds in an incremental way.

Including across profiles?
And there's certainly no point in downloading the various `make` based builds.

**Review by SkySkimmer — APPROVED** [review]

File: .gitlab-ci.yml
Comment (ejgallego) [reply .gitlab-ci.yml]: Builds live in "contexts", the only one is default but you for example could declare one for `bisect_pxx` etc... they are independent.

`(env ...)` profiles are supposed to be tracked in the dependency chain, so indeed if you change some stuff or profile it may trigger a rebuild. Note that this is still being developed AFAICT.

Comment (ejgallego) [reply .gitlab-ci.yml]: > And there's certainly no point in downloading the various make based builds.

Sure, I was just mentioning in case in the future we need it. The dependency tracking is more useful for the developer tho.

**Pull request merged**
